src/ai/defaultai.h

```diff
@@ -375,10 +380,16 @@ struct DefaultAI : ComputerPlayer {
 	// id of iron as resource to identify iron mines in mines_per_type map
 	int32_t iron_resource_id = Widelands::INVALID_INDEX;
 
+	// Record with immovable attribute is created or collected by which building names and attributes
+	std::map<Widelands::MapObjectDescr::AttributeIndex, std::set<ImmovableAttribute>>
+	   buildings_immovable_attributes_;
```

Comment:
I must say the variable name is difficult to understand (what it is about| also description above is not very helpful.

Reply:
ok I will try to find a better name then.

Reply:
It would be good. I am trying to read the code and all the time I just stuck and try to guess what a variable can contain. 
When I look at ImmovableAttribute, then building_immovable_attributes is something like:
<Building index: <Building name : Building Attribute>> 
correct?

Reply:
Well, I admit I had difficulties to understand all of this when I tried to understand the changes of gunchleoc. but as my changes are about 2 month old (hoped she could incorporate them by herself) I need to dig into this by myself again. 
I believe to remember this is an index (a set) of all immovable attributes that a building is interacting with.

Reply:
Because building name vs building index is 1:1, so this does not bring any new information to have both. But I can admit that there may be reasons that this approach is the best

Reply:
> 
> When I look at ImmovableAttribute, then building_immovable_attributes is something like:
> <Building index: <Building name : Building Attribute>>
> correct?

No. 
Instead it is 
<AttributeIndex: <Building name : Building Attribute>>
where Attribute index is the index of an immovable attribute that is created, collected or needed by the building. eg the index of the attribute "tree" or the attribute "pond_dry"
Building Attribute is something like kNeedsRocks

Reply:
How about renaming it to buildings_per_immovable_attributes_ as it lists the buildings (and their building attributes) for each immovable attribute they interact with

Reply:
the word "attribute" seems not good to me. It mean ~'map object', 'collectable object' or 'surface resource', doesnt it? Tings like "colected_attributes" might be renamed to collected_map_objects or so.

Also do we need Building Attribute in buildings_immovable_attributes_, because it "attribute" and "building attribute" refers to the same things berry <-> kNeedBerry.

What about "collectable_to_collectors_map" or "buildings_per_collected_objects" or similar.

I know that it is only cosmetic but it would increase readibiity of code IMHO.

Reply:
for the first question we need the building attribute and the name as with these changes we have multiple rangers (currently only for the amazons: jungle_preserver and rare_tree_plantation) and we can have multiple other types theoretically (currently a warning is thrown if we have multiple of them) so sometimes in the code we use the name and sometimes the attribute.

having collect in the variable name isn't recommendable as not only collected immovables are listed but also created immovables (i.e. we have immovables with the attribute sapling where the normal ranger or the amazon rare tree plantationb is assigned, we have the attrib pole where amazon ranger is assigned, ) furthermore even immovables that are only needed and neither collected or created are to be recorded there (eg. the amazon wilderness_keeper needs a tree to free game)

Reply:
I see that so much stuff was changed since I worked on AI, I would perhaps need to put it on paper to understand it properly. But makes more sense now when you explained it to me, thanks

Reply:
no worries. take your time. However all I tried in this branch was fixing some logical errors in GunChleocs attempt to read some former AI hints from the production programs and make them work properly for some particular changes it might be worth reading my comments on the original branch (#4465)